### PR TITLE
[bazel] use GPU machine and run GPU tests

### LINKED ISSUE #79354: [CI] [bazel] run GPU-requiring tests on a GPU machine
### 🚀 The feature, motivation and pitch

At the moment we use the same CPU-only build machine for tests in bazel build.
This prevents GPU-requiring tests from being tested in CI.

There are couple of ways to go about it:
1. If https://github.com/pytorch/pytorch/issues/79348 could be addressed and the time could be dropped to 15 minutes, we can use simply use the same machine with GPU for both build and test.
2. Otherwise, we would need to have a step that packages pytorch artifacts, including tests and ships it to a machine where we can run it without requiring bazel (having bazel in the loop could lead to costly re-compilation, so it should be intentionally left out of the loop).

### Alternatives

Keep not testing GPU tests with bazel in CI

### Additional context

_No response_

cc @seemethere @malfet @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by vors**
Fixes #79354

Run bazel build and tests on GPU machine.

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Is there any reason not to split this test in two similar to the other ones we have: one for CPU on a CPU only machine and one for GPU on a GPU machine?

File: .github/workflows/_bazel-build-test.yml [hunk 1]
```diff
@@ -26,7 +26,7 @@ jobs:
   build-and-test:
     # Don't run on forked repos.
     if: github.repository_owner == 'pytorch'
-    runs-on: [self-hosted, linux.4xlarge]
+    runs-on: [self-hosted, linux.4xlarge.nvidia.gpu]
```
Comment (huydhn) [inline .github/workflows/_bazel-build-test.yml:29]: I'm a bit confuse about the other https://github.com/pytorch/pytorch/pull/95795 to have cpu support, is that what we current have already?  IMO, let's just keep it in this PR

On the other hand, your approach of creating one job for cpu while other for gpu makes sense to me.  So pull workflow would look as follows:

I suggest to use `test-matrix` here to keep the same syntax as other jobs

```
linux-bionic-cpu-py3_10-gcc7-bazel-test:
  name: linux-bionic-cpu-py3.10-gcc7-bazel-test
  uses: ./.github/workflows/_bazel-build-test.yml
  with:
    build-environment: linux-bionic-cuda11.7-py3.10-gcc7-bazel-test
    docker-image-name: pytorch-linux-bionic-cuda11.7-cudnn8-py3-gcc7
    cuda-version: cpu
    test-matrix: |
      { include: [
        { config: "default", shard: 1, num_shards: 1, runner: "linux.4xlarge" },
      ]}

linux-bionic-cuda11_7-py3_10-gcc7-bazel-test:
  name: linux-bionic-cuda11.7-py3.10-gcc7-bazel-test
  uses: ./.github/workflows/_bazel-build-test.yml
  with:
    build-environment: linux-bionic-cuda11.7-py3.10-gcc7-bazel-test
    docker-image-name: pytorch-linux-bionic-cuda11.7-cudnn8-py3-gcc7
    test-matrix: |
      { include: [
        { config: "default", shard: 1, num_shards: 1, runner: "linux.4xlarge.nvidia.gpu" },
      ]}
```

File: .github/workflows/_bazel-build-test.yml [hunk 2]
```diff
@@ -11,13 +11,25 @@ on:
...
         default: ""
         description: |
           If this is set, our linter will use this to make sure that every other
           job with the same `sync-tag` is identical.
+      test-matrix:
+        required: false
+        type: string
+        description: |
+          An option JSON description of what test configs to run later on. This
+          is moved here from the Linux test workflow so that we can apply filter
```
Comment (albanD) [inline .github/workflows/_bazel-build-test.yml:31]: nit: this second sentence can be updated here: just say that we provide all test configs at build time to be able to only do the necessary work.

**Review by albanD — COMMENTED** [review]
The cuda-version arg is a bit confusing as it doesn't refer to an actual cuda version.
But that sounds good to me as a first iteration. I'll let @huydhn take a look as well.

File: .github/workflows/_bazel-build-test.yml [hunk 3]
```diff
@@ -11,13 +11,25 @@ on:
...
+        default: "default"
       sync-tag:
         required: false
         type: string
         default: ""
         description: |
           If this is set, our linter will use this to make sure that every other
           job with the same `sync-tag` is identical.
+      test-matrix:
+        required: false
```
Comment (huydhn) [inline .github/workflows/_bazel-build-test.yml:27]:
```suggestion
        required: true
```

File: .github/workflows/_bazel-build-test.yml [hunk 4]
```diff
@@ -11,13 +11,25 @@ on:
...
+        type: string
+        description: |
```
Comment (huydhn) [inline .github/workflows/_bazel-build-test.yml:29]:
```suggestion
        description: | 
          A JSON description of what configs to run later on.
```

File: .github/workflows/_bazel-build-test.yml [hunk 5]
```diff
@@ -71,6 +83,14 @@ jobs:
         with:
           github-token: ${{ secrets.GITHUB_TOKEN }}
 
+      # Apply the filter logic to the build step too if the test-config label is already there
```
Comment (huydhn) [inline .github/workflows/_bazel-build-test.yml:86]: You don't need to worry about all this `filter-test-configs` step in this PR.  It's not needed and can be removed

File: .github/workflows/_bazel-build-test.yml [hunk 6]
```diff
@@ -26,7 +38,7 @@ jobs:
   build-and-test:
     # Don't run on forked repos.
     if: github.repository_owner == 'pytorch'
-    runs-on: [self-hosted, linux.4xlarge]
+    runs-on: ${{ inputs.runner-type }}
```
Comment (huydhn) [inline .github/workflows/_bazel-build-test.yml:41]: Here is how the `test-matrix` should be used here

```
    strategy:
      matrix: ${{ fromJSON(inputs.test-matrix) }}
      fail-fast: false
    runs-on: ${{ matrix.runner }}
```

File: .github/workflows/_bazel-build-test.yml [hunk 7]
```diff
@@ -11,13 +11,25 @@ on:
         required: true
         type: string
         description: Name of the base docker image to build with.
+      cuda-version:
```
Comment (huydhn) [inline .github/workflows/_bazel-build-test.yml:14]: I guess you find the `cuda-version` param from windows CI, that's probably ok I think

File: .github/workflows/_bazel-build-test.yml [hunk 8]
```diff
@@ -11,13 +11,25 @@ on:
...
+        required: false
+        type: string
+        description: What CUDA version to build with, "cpu" for none.
+        default: "default"
```
Comment (huydhn) [inline .github/workflows/_bazel-build-test.yml:18]:
```suggestion
        default: "cpu"
```

File: .github/workflows/pull.yml [hunk 9]
```diff
@@ -294,12 +294,28 @@ jobs:
...
+          { config: "default", shard: 1, num_shards: 1, runner: "linux.4xlarge" },
+        ]}
+
   linux-bionic-cuda11_7-py3_10-gcc7-bazel-test:
     name: linux-bionic-cuda11.7-py3.10-gcc7-bazel-test
     uses: ./.github/workflows/_bazel-build-test.yml
     with:
       build-environment: linux-bionic-cuda11.7-py3.10-gcc7-bazel-test
       docker-image-name: pytorch-linux-bionic-cuda11.7-cudnn8-py3-gcc7
+      test-matrix: |
```
Comment (huydhn) [inline .github/workflows/pull.yml:315]: Add `cuda-version: "11.7"`?  I know that this is not used internally (yet), but it makes it clearer I think.

**Review by huydhn — APPROVED** [review]
LGTM!  Thank you very much for extending bazel build and test to GPU

**Pull request closed without merging**